### PR TITLE
feat: generate build configs [AR-1463]

### PR DESCRIPTION
**Pull request by MohammedMokresh**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-1463" title="AR-1463" target="_blank">TaskAR-1463</a>  Build config flags
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

To make kalium aware of the build configs we needed to generate and update the core logic with the generated values

### Solutions

generated build configs from the default.json and add implementation for a custom remote repositories for a custom config builds 
then pass those generated values to kalium

### Dependencies

Needs releases with:

- https://github.com/wireapp/kalium/pull/557

#### How to Test

build the app and check the BuildConfig if it was updated with the values from default.json
check the getBuildConfigUseCase if it's returning the generated values 

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/1/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3511865)(3511865)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3511865)(3511865)-dev-release.apk)

**Review by vitorhugods — CHANGES_REQUESTED** [review]
I like it a lot! 🥰
Really well done and clean.

I guess a lot of these feature flags will go to space for now, but at least we have a way cleaner way of doing this, compared to current Scala app.

Just some suggestions to make it tighter and a big reminder that this should not be merged before https://github.com/wireapp/kalium/pull/557.

I updated the PR description to mention the blocking PR.

File: buildSrc/src/main/kotlin/Configs.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    CERTIFICATE_PIN("certificatePin", ConfigType.CERTIFICATE_PIN),
+
+
+    CANDIDATE("candidate", ConfigType.FLAVOUR_CONFIG),
+    EXPERIMENTAL("experimental", ConfigType.FLAVOUR_CONFIG),
+    FDROID("fdroid", ConfigType.FLAVOUR_CONFIG),
+    DEV("dev", ConfigType.FLAVOUR_CONFIG),
+    INTERNAL("internal", ConfigType.FLAVOUR_CONFIG),
+    PROD("prod", ConfigType.FLAVOUR_CONFIG);
```
Comment (ohassine) [inline buildSrc/src/main/kotlin/Configs.kt:55]: And extract this to VariantConfigs ?

File: buildSrc/src/main/kotlin/scripts/variants.gradle.kts [hunk 2]
```diff
@@ -107,5 +119,68 @@ android {
...
+                ConfigType.STRING -> {
+                    flavor.buildConfigField(
+                        configs.configType.type, configs.name, """"${buildtimeConfiguration?.configuration?.get(configs.value)}""""
+                    )
+                }
+                ConfigType.INT, ConfigType.BOOLEAN -> {
+                    flavor.buildConfigField(
+                        configs.configType.type, configs.name, buildtimeConfiguration?.configuration?.get(configs.value).toString()
+                    )
+                }
```
Comment (ohassine) [inline buildSrc/src/main/kotlin/scripts/variants.gradle.kts:134]: Isn't the same thing, can't we do for both: 
`""""${buildtimeConfiguration?.configuration?.get(configs.value).toString()}""""`

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3519155)(3519155)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3519155)(3519155)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/di/KaliumConfigsModule.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            httpProxyUrl = BuildConfig.HTTP_PROXY_URL,
+            supportEmail = BuildConfig.SUPPORT_EMAIL,
+            teamsUrl = BuildConfig.TEAMS_URL,
+            backendUrl = BuildConfig.BACKEND_URL,
+            accountUrl = BuildConfig.ACCOUNT_URL,
+            maxAccount = BuildConfig.MAX_ACCOUNTS,
+            newPasswordMaximumLength = BuildConfig.NEW_PASSWORD_MAXIMUM_LENGTH,
+            newPasswordMinimumLength = BuildConfig.NEW_PASSWORD_MINIMUM_LENGTH,
+            passwordMaxAttempts = BuildConfig.PASSWORD_MAX_ATTEMPTS,
+            appLockTimeout = BuildConfig.APP_LOCK_TIMEOUT
```
[the remark this answers is not in the document]
Comment (MohammedMokresh) [reply app/src/main/kotlin/com/wire/android/di/KaliumConfigsModule.kt:53]: yeah for the urls, I spoke with Mohamad , and I will update it after he merge the api versioning 
or he will update this if I merged first

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/6/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3519434)(3519434)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3519434)(3519434)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/7/) **failed**.

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/8/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3527364)(3527364)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3527364)(3527364)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/9/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3528077)(3528077)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3528077)(3528077)-dev-release.apk)

**Review by ohassine — APPROVED** [review]
lgtm 🚀

**Review by yamilmedina — APPROVED** [review]
Great work Big Mo! nice work in gradle world 👏

**Comment by AndroidBob** [comment]
Build [10](https://10.10.124.134/job/wire-android-reloaded/job/PR-636/10/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3572468)(3572468)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3572468)(3572468)-dev-release.apk)

**Pull request merged**